File: backend/src/entities/StorageLocation.js
import { EntitySchema } from '@mikro-orm/core';

/**
 * Default storage locations created for every new pantry
 * @type {Array<{name: string, icon: string, order: number}>}
 */
export const DEFAULT_STORAGE_LOCATIONS = [
  { name: 'Fridge', icon: 'kitchen', order: 0 },
  { name: 'Freezer', icon: 'ac_unit', order: 1 },
  { name: 'Cupboard', icon: 'shelves', order: 2 },
  { name: 'Cellar', icon: 'warehouse', order: 3 },
  { name: 'Other', icon: 'inventory_2', order: 4 },
];

/**
 * @typedef {Object} StorageLocationProperties
 * @property {string} id - UUID primary key
 * @property {string} name - Location name (e.g. Fridge, Freezer)
 * @property {string} icon - Material icon name
 * @property {number} order - Display order
 * @property {Date} createdAt - Creation timestamp
 * @property {Date} updatedAt - Last update timestamp
 */

/**
 * Storage Location Entity Schema
 * @type {EntitySchema<StorageLocationProperties>}
 */
export const StorageLocationSchema = new EntitySchema({
  name: 'StorageLocation',
  tableName: 'storage_location',
  properties: {
    id: {
      type: 'uuid',
      primary: true,
      onCreate: () => crypto.randomUUID(),
    },
    pantry: {
      kind: 'm:1',
      entity: 'Pantry',
    },
    name: {
      type: 'string',
    },
    icon: {
      type: 'string',
      default: 'inventory_2',
    },
    order: {
      type: 'number',
      default: 0,
    },
    createdAt: {
      type: 'Date',
      onCreate: () => new Date(),
    },
    updatedAt: {
      type: 'Date',
      onCreate: () => new Date(),
      onUpdate: () => new Date(),
    },
  },
});

File: mobile/lib/models/storage_location.dart
import 'package:flutter/material.dart';

class StorageLocation {
  final String id;
  final String name;
  final String icon;
  final int order;
  final DateTime createdAt;
  final DateTime updatedAt;

  StorageLocation({
    required this.id,
    required this.name,
    this.icon = 'inventory_2',
    this.order = 0,
    required this.createdAt,
    required this.updatedAt,
  });

  factory StorageLocation.fromJson(Map<String, dynamic> json) {
    return StorageLocation(
      id: json['id'] as String,
      name: json['name'] as String,
      icon: json['icon'] as String? ?? 'inventory_2',
      order: json['order'] as int? ?? 0,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'icon': icon,
      'order': order,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  StorageLocation copyWith({
    String? id,
    String? name,
    String? icon,
    int? order,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return StorageLocation(
      id: id ?? this.id,
      name: name ?? this.name,
      icon: icon ?? this.icon,
      order: order ?? this.order,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  /// Map icon string names to Material Icons
  IconData get iconData {
    switch (icon) {
      case 'kitchen':
        return Icons.kitchen;
      case 'ac_unit':
        return Icons.ac_unit;
      case 'shelves':
        return Icons.shelves;
      case 'warehouse':
        return Icons.warehouse;
      case 'inventory_2':
        return Icons.inventory_2;
      default:
        return Icons.inventory_2;
    }
  }
}
